src/Data/Tree/Binary.agda

```diff
@@ -10,56 +10,70 @@ module Data.Tree.Binary where
 
 open import Level using (Level; _⊔_)
 open import Data.List.Base using (List)
-open import Data.DifferenceList as DiffList using (DiffList; []; _∷_; _∷ʳ_; _++_)
+open import Data.DifferenceList as DiffList using (DiffList; []; _∷_; _∷ʳ_; _++_; [_])
 open import Data.Nat.Base using (ℕ; zero; suc; _+_)
 open import Function.Base
-open import Relation.Binary.PropositionalEquality
 
 private
   variable
-    a b : Level
+    a b c d : Level
     A : Set a
     B : Set b
+    C : Set c
+    D : Set d
 
-data Tree (A : Set a) : Set a where
-  leaf : Tree A
-  node : Tree A → A → Tree A → Tree A
+data Tree (A : Set a) (B : Set b) : Set (a ⊔ b) where
+  leaf : B → Tree A B
+  node : Tree A B → A → Tree A B → Tree A B
 
-map : (A → B) → Tree A → Tree B
-map f leaf         = leaf
-map f (node l m r) = node (map f l) (f m) (map f r)
+map : (A → C) → (B → D) → Tree A B → Tree C D
+map f g (leaf x)     = leaf (g x)
+map f g (node l m r) = node (map f g l) (f m) (map f g r)
 
-size : Tree A → ℕ
-size leaf         = 0
+size : Tree A B → ℕ
+size (leaf x)     = 0
 size (node l m r) = size l + suc (size r)
```

Comment:
I think this should be renamed to `nodes` for consistency with the other notions
of size one being called `leaves`. Or rather, it may be a good idea to call it
something like `#nodes` while `nodes` would be a function `Tree A B -> List A`!